src/pudl/extract/eia_bulk_elec.py

```diff
@@ -0,0 +1,104 @@
+"""Module to extract aggregate data from the EIA bulk electricity download.
+
+EIA's bulk electricity data contains 680,000 timeseries. These timeseries contain a
+variety of measures (fuel amount and cost are just two) across multiple levels of
+aggregation, from individual plants to national averages.
+
+The data is formatted as a single 1.1GB text file of line-delimited JSON with one line
+per timeseries. Each JSON structure has two nested levels: the top level contains
+metadata describing the series and the second level (under the "data" heading)
+contains an array of timestamp/value pairs. This structure leads to a natural
+normalization into two tables: one of metadata and one of timeseries. That is the
+format delivered by this module.
+"""
+from io import BytesIO
+from pathlib import Path
+
+import pandas as pd
+
+from pudl.workspace.datastore import Datastore
+
+
+def _filter_for_fuel_receipts_costs_series(df: pd.DataFrame) -> pd.DataFrame:
+    """Pick out the desired data series."""
+    series_to_extract = ["RECEIPTS_BTU", "COST_BTU"]
+    series_pattern = "|".join(series_to_extract)
+    pattern = rf"^ELEC\.(?:{series_pattern})"
+    try:
+        _filter = df.loc[:, "series_id"].str.match(pattern, na=False)
+        return df.loc[_filter, :]
+    except KeyError:
+        return pd.DataFrame()  # empty
+
+
+def _filter_and_read_to_dataframe(raw_zipfile: Path) -> pd.DataFrame:
+    """Decompress and filter the 1100 MB file down to the 16 MB we actually want.
+
+    This produces a dataframe with all text fields. The timeseries data is
+    left as JSON strings in the 'data' column. The other columns are metadata.
+    """
+    filtered = []
+    # Use chunksize arg to reduce peak memory usage when reading in 1.1 GB file
+    # For reference, the file has ~680k lines and we want around 8.5k
+    with pd.read_json(
+        raw_zipfile, compression="zip", lines=True, chunksize=10_000
+    ) as reader:
+        for chunk in reader:
+            filtered.append(_filter_for_fuel_receipts_costs_series(chunk))
+    out = pd.concat(filtered, ignore_index=True)
+    return out
+
+
+def _parse_data_column(elec_df: pd.DataFrame) -> pd.DataFrame:
+    out = []
+    for idx in elec_df.index:
```

Comment:
Is this converting every individual row into its own dataframe, and then parsing just that row? Is that not annoyingly slow? Could it be broken into 3 large dataframes instead, one for each type of date, that are each processed on their own and then concatenated together?

Reply:
It's a nested data structure; a dataframe where one column contains JSON strings that each need conversion to dataframes. I think that has to be traversed row-wise? I guess I could use `df.iteritems()` or `df.apply()` instead, or totally rewrite the extract around `pd.json_normalize()` but I don't think it would make much performance difference. By far the biggest time suck of this ETL is simply the initial decompression and deserialization of the big zip file (90 seconds on my old desktop).

Reply:
Okay that makes sense. I think some explanation of the data structures in the docstrings of these functions (even when they are `_private`) would help make it clearer why it's being done this way.